### PR TITLE
Fix delete branch perm checking

### PR DESCRIPTION
**Pull request by lunny**

### CODE REVIEW

File: services/repository/branch.go [hunk 1]
```diff
@@ -463,15 +466,21 @@ var (
 	ErrBranchIsDefault = errors.New("branch is default")
 )
 
-// DeleteBranch delete branch
-func DeleteBranch(ctx context.Context, doer *user_model.User, repo *repo_model.Repository, gitRepo *git.Repository, branchName string) error {
-	err := repo.MustNotBeArchived()
+func CanDeleteBranch(ctx context.Context, repo *repo_model.Repository, branchName string, doer *user_model.User) error {
```
Comment (hiifong) [inline services/repository/branch.go:468]: ~~Maybe it would be better to return a boolean type~~

Comment (lunny) [reply services/repository/branch.go:468]: It returns an error so that we know the reason why it can not delete branch.

File: routers/web/repo/pull.go [hunk 2]
```diff
@@ -1435,13 +1435,12 @@ func CleanUpPullRequest(ctx *context.Context) {
...
-		ctx.ServerError("GetUserRepoPermission", err)
-		return
-	}
-	if !perm.CanWrite(unit.TypeCode) {
-		ctx.NotFound("CleanUpPullRequest", nil)
+	if err := repo_service.CanDeleteBranch(ctx, pr.HeadRepo, pr.HeadBranch, ctx.Doer); err != nil {
+		if errors.Is(err, access_model.ErrNoPermission{}) {
+			ctx.NotFound("CleanUpPullRequest", nil)
+		} else {
+			ctx.ServerError("GetUserRepoPermission", err)
```
Comment (Zettat123) [inline routers/web/repo/pull.go]:
```suggestion
			ctx.ServerError("CanDeleteBranch", err)
```

File: routers/api/v1/repo/pull.go [hunk 3]
```diff
@@ -1057,49 +1057,53 @@ func MergePullRequest(ctx *context.APIContext) {
...
-
-		var headRepo *git.Repository
-		if ctx.Repo != nil && ctx.Repo.Repository != nil && ctx.Repo.Repository.ID == pr.HeadRepoID && ctx.Repo.GitRepo != nil {
-			headRepo = ctx.Repo.GitRepo
-		} else {
-			headRepo, err = gitrepo.OpenRepository(ctx, pr.HeadRepo)
+	if form.DeleteBranchAfterMerge && pr.Flow == issues_model.PullRequestFlowGithub {
+		// check permission even it has been checked in repo_service.DeleteBranch so that we don't need to
+		// do RetargetChildrenOnMerge
+		if err := repo_service.CanDeleteBranch(ctx, pr.HeadRepo, pr.HeadBranch, ctx.Doer); err == nil {
```
Comment (wxiaoguang) [inline routers/api/v1/repo/pull.go:1064]: I do not think the check is necessary or right.

image

File: routers/web/repo/pull.go [hunk 4]
```diff
@@ -1435,13 +1435,12 @@ func CleanUpPullRequest(ctx *context.Context) {
...
 
-	perm, err := access_model.GetUserRepoPermission(ctx, pr.HeadRepo, ctx.Doer)
-	if err != nil {
-		ctx.ServerError("GetUserRepoPermission", err)
-		return
-	}
-	if !perm.CanWrite(unit.TypeCode) {
-		ctx.NotFound("CleanUpPullRequest", nil)
+	if err := repo_service.CanDeleteBranch(ctx, pr.HeadRepo, pr.HeadBranch, ctx.Doer); err != nil {
+		if errors.Is(err, access_model.ErrNoPermission{}) {
```
Comment (wxiaoguang) [inline routers/web/repo/pull.go]: It is always false

File: routers/web/repo/pull.go (hunk 2 above)
Comment (lunny) [reply routers/web/repo/pull.go]: b3582d0

File: routers/web/repo/pull.go (hunk 4 above)
Comment (lunny) [reply routers/web/repo/pull.go]: b3582d0b2e7b2dce12a09e2a5fe0a9b3de13b2cd

File: routers/api/v1/repo/pull.go (hunk 3 above)
Comment (lunny) [reply routers/api/v1/repo/pull.go:1064]: The retarget actions will run before deleting the branches. If we don't check whether the branch can be deleted, maybe the retarget actions succeed, but the deleting branch fails.

Comment (wxiaoguang) [reply routers/api/v1/repo/pull.go:1064]: `RetargetChildrenOnMerge`  already checks `pr.BaseRepoID == pr.HeadRepoID`, then in the same repo, the merger should also be able to delete?

Comment (wxiaoguang) [reply routers/api/v1/repo/pull.go:1064]: And TBH I don't quite understand the logic, so no idea whether it is right.

Comment (lunny) [reply routers/api/v1/repo/pull.go:1064]: > `RetargetChildrenOnMerge` already checks `pr.BaseRepoID == pr.HeadRepoID`, then in the same repo, the merger should also be able to delete?

I also noticed this logic and I don't think it's right, why only the same repository? But I will only fix the bug here and leave the refactor to other pull requests.

File: routers/api/v1/repo/branch.go [hunk 5]
```diff
@@ -150,11 +150,6 @@ func DeleteBranch(ctx *context.APIContext) {
 		}
 	}
 
-	if ctx.Repo.Repository.IsMirror {
-		ctx.Error(http.StatusForbidden, "IsMirrored", fmt.Errorf("can not delete branch of an mirror repository"))
-		return
-	}
-
```
Comment (wxiaoguang) [inline routers/api/v1/repo/branch.go:157]: Related or unrelated change?

Comment (lunny) [reply routers/api/v1/repo/branch.go:157]: It's an unrelated change. But it's a duplicated line as line 129. But it's so nit, I would like to merge it into this PR.

File: models/perm/access/repo_permission.go [hunk 6]
```diff
@@ -18,6 +18,21 @@ import (
...
 )
 
+type ErrPermissionDenied struct {
+	RepoID int64
+	Unit   unit.Type
+	Perm   perm_model.AccessMode
+}
+
+func (e ErrPermissionDenied) Error() string {
+	return fmt.Sprintf("permission denied to access repo %d unit %s with mode %s", e.RepoID, e.Unit.LogString(), e.Perm.LogString())
```
Comment (wxiaoguang) [inline models/perm/access/repo_permission.go]: It could use `Errorf("..., %w", util.ErrPermissionDenied)`

Then could check it by `errors.Is(err, util.ErrPermissionDenied)`

Comment (wxiaoguang) [reply models/perm/access/repo_permission.go]: Or remove this new error completely.

Just use `util.NewPermissionDeniedErrorf(".....")`

Comment (lunny) [reply models/perm/access/repo_permission.go]: cf1d71a8b6

**Review by wxiaoguang — APPROVED** [review]
Still, I think:

1. it needs some tests
2. it needs to clarify why checking "pr.Flow == issues_model.PullRequestFlowGithub" (comments)

File: routers/api/v1/repo/pull.go [hunk 7]
```diff
@@ -1057,49 +1057,54 @@ func MergePullRequest(ctx *context.APIContext) {
...
-				ctx.Error(http.StatusForbidden, "IsProtectedBranch", fmt.Errorf("branch protected"))
-			default:
-				ctx.Error(http.StatusInternalServerError, "DeleteBranch", err)
+			if exist {
+				ctx.Status(http.StatusOK)
+				return
+			}
+
+			var headRepo *git.Repository
+			if ctx.Repo != nil && ctx.Repo.Repository != nil && ctx.Repo.Repository.ID == pr.HeadRepoID && ctx.Repo.GitRepo != nil {
```
Comment (hiifong) [inline routers/api/v1/repo/pull.go:1077]: Maybe, we need a way to check if repo is nil， What do you think? @wxiaoguang @lunny @Zettat123

**Review by hiifong — APPROVED** [review]

File: routers/api/v1/repo/pull.go (hunk 7 above)
Comment (lunny) [reply routers/api/v1/repo/pull.go:1077]: Maybe we can create a method for `ctx.Repo` in future PRs.
